packages/components/form/src/form.vue

```diff
@@ -1,258 +1,158 @@
-          ? fields.filter((field) => props === field.prop)
-          : fields.filter((field) => props.indexOf(field.prop) > -1)
-        : fields
-      fds.forEach((field) => {
-        field.clearValidate()
-      })
+    if (fieldsError) {
+      valid = false
+      if (!firstInvalidFields) firstInvalidFields = fieldsError
     }
 
-    const validate = (callback?: Callback) => {
-      if (!props.model) {
-        debugWarn('Form', 'model is required for validate to work!')
-        return
-      }
-
-      let promise: Promise<boolean> | undefined
-      // if no callback, return promise
-      if (typeof callback !== 'function') {
-        promise = new Promise((resolve, reject) => {
-          callback = function (valid, invalidFields) {
-            if (valid) {
-              resolve(true)
-            } else {
-              reject(invalidFields)
-            }
-          }
-        })
-      }
-
-      if (fields.length === 0) {
-        callback(true)
-      }
-      let valid = true
-      let count = 0
-      let invalidFields = {}
-      let firstInvalidFields
-      for (const field of fields) {
-        field.validate('', (message, field) => {
-          if (message) {
-            valid = false
-            firstInvalidFields || (firstInvalidFields = field)
-          }
-          invalidFields = { ...invalidFields, ...field }
-          if (++count === fields.length) {
-            callback(valid, invalidFields)
-          }
-        })
-      }
-      if (!valid && props.scrollToError) {
-        scrollToField(Object.keys(firstInvalidFields)[0])
-      }
-      return promise
-    }
+    invalidFields = { ...invalidFields, ...fieldsError }
+  }
 
-    const validateField = (
-      props: string | string[],
-      cb: ValidateFieldCallback
-    ) => {
-      props = [].concat(props)
-      const fds = fields.filter((field) => props.indexOf(field.prop) !== -1)
-      if (!fields.length) {
-        debugWarn('Form', 'please pass correct props!')
-        return
-      }
+  if (!valid) {
+    if (scrollToError) scrollToField(Object.keys(firstInvalidFields!)[0])
+    return Promise.reject(invalidFields)
+  }
+}
 
-      fds.forEach((field) => {
-        field.validate('', cb)
-      })
-    }
+const scrollToField = (prop: string) => {
+  const field = filterFields(fields, prop)[0]
+  if (field) {
+    field.$el?.scrollIntoView()
+  }
+}
 
-    const scrollToField = (prop: string) => {
-      fields.forEach((item) => {
-        if (item.prop === prop) {
-          item.$el.scrollIntoView?.()
-        }
-      })
-    }
+watch(
+  () => props.rules,
+  () => {
+    if (props.validateOnRuleChange) validate()
+  }
+)
 
-    const elForm = reactive({
-      ...toRefs(props),
-      resetFields,
-      clearValidate,
-      validateField,
-      emit,
-      addField,
-      removeField,
-      ...useFormLabelWidth(),
-    })
+provide(
+  formContextKey,
+  reactive({
+    ...toRefs(props),
+    emit,
```

Comment:
**Question:** what do you think about change this to `formEmit` to avoid naming issue?

Reply:
It also removes confusion here.

Reply:
Since it is `formContextKey`, it must be related to `form`, and I think there is no need to repeat the name.

Reply:
The reason I commented on this one is because that this is provided by `form` but it was consumed by `form-item` so that we should be considering the downstream in this case `form-item`, `form-item` has its own `emit` method, and we provides `form`'s emit, under this circumstance we should consider renaming the method or we provide a delegated method something like `emitsToParent`

Reply:
Is it really necessary? The variable name here is already explicitly `form`.
```ts
formContext.emit('validate', props.prop, !errors, validateMessage.value)
```

Reply:
Well, since in the consumer side we are using it like `formContext.emit` I think this wouldn't be a problem, resolving this